project_euler/problem_107/sol1.py

```diff
@@ -0,0 +1,131 @@
+"""
+The following undirected network consists of seven vertices and twelve edges
+with a total weight of 243.
+￼
+The same network can be represented by the matrix below.
+
+    A   B   C   D   E   F   G
+A   -   16  12  21  -   -   -
+B   16  -   -   17  20  -   -
+C   12  -   -   28  -   31  -
+D   21  17  28  -   18  19  23
+E   -   20  -   18  -   -   11
+F   -   -   31  19  -   -   27
+G   -   -   -   23  11  27  -
+
+However, it is possible to optimise the network by removing some edges and still
+ensure that all points on the network remain connected. The network which achieves
+the maximum saving is shown below. It has a weight of 93, representing a saving of
+243 - 93 = 150 from the original network.
+
+Using network.txt (right click and 'Save Link/Target As...'), a 6K text file
+containing a network with forty vertices, and given in matrix form, find the maximum
+saving which can be achieved by removing redundant edges whilst ensuring that the
+network remains connected.
+"""
+
+import os
+from typing import Dict, List, Set, Tuple
+
+Edge = Tuple[int, int]
+
+
+class Graph:
+    """
+    A class representing an undirected weighted graph.
+    """
+
+    def __init__(self, vertices: Set[int], edges: Dict[Edge, int]) -> None:
+        """
+        >>> graph = Graph({3,4,1,2},{(2,1):5, (1,3):10, (4,1):20, (2,4):30, (4,3):1})
+        >>> sorted(graph.vertices)
+        [1, 2, 3, 4]
+        >>> sorted([(v,k) for k,v in graph.edges.items()])
+        [(1, (3, 4)), (5, (1, 2)), (10, (1, 3)), (20, (1, 4)), (30, (2, 4))]
+        """
+        self.vertices: Set[int] = vertices
+        self.edges: Dict[Edge, int] = {
+            (min(edge), max(edge)): weight for edge, weight in edges.items()
+        }
+
+    def add_edge(self, edge: Edge, weight: int) -> None:
+        """
+        >>> graph = Graph({1, 2}, {(2, 1): 4})
+        >>> graph.add_edge((3, 1), 5)
+        >>> sorted(graph.vertices)
+        [1, 2, 3]
+        >>> sorted([(v,k) for k,v in graph.edges.items()])
+        [(4, (1, 2)), (5, (1, 3))]
+        """
+        self.vertices.add(edge[0])
+        self.vertices.add(edge[1])
+        self.edges[(min(edge), max(edge))] = weight
+
+
+def prims_algorithm(graph: Graph) -> Graph:
+    """
+    Run Prim's algorithm on a Graph to find the minimum spanning tree.
+    Reference: https://en.wikipedia.org/wiki/Prim%27s_algorithm
+    >>> graph = Graph({1,2,3,4},{(1,2):5, (1,3):10, (1,4):20, (2,4):30, (3,4):1})
+    >>> mst = prims_algorithm(graph)
+    >>> sorted(mst.vertices)
+    [1, 2, 3, 4]
+    >>> sorted(mst.edges)
+    [(1, 2), (1, 3), (3, 4)]
+    """
+    subgraph: Graph = Graph({min(graph.vertices)}, {})
+    min_edge: Edge
+    min_weight: int
+    edge: Edge
+    weight: int
```

Comment:
This is just a suggestion and you don't have to make any changes:

There's no need to provide a type hint for every variable as whichever static type checker you are using will probably be able to infer the variable type according to the assignment. In simple cases like these, it is not required.